core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/DataProcessor.scala

```diff
@@ -194,7 +194,13 @@ class DataProcessor( // dependencies:
   }
 
   def shutdown(): Unit = {
-    dpThread.cancel(true)
+    if (stateManager.confirmState(Running)) {
+      pauseManager.pause().onSuccess { ret =>
+        dpThread.cancel(true)
```

Comment:
A few comments:
1. we should call operator.close() properly, otherwise we will have resource leaks issues, especially for source and sink operators. I checked our code, and there is currently zero place that we call close().
2. I'm wondering if shutdown should also returns a future to indicate the success of a graceful shutdown

Reply:
1. agreed, actually, I forgot we have a close() function to call.
2. for now shutdown() will only be called when the worker kills itself. I don't know if that future makes sense. If so, who should use that future?